### ISSUE TITLE
Decreasing the number of triangles when doing triangulation

### ISSUE THREAD

**Issue by swy0930** [xBimTeam/XbimWebUI#67]
Hello,

The wexBIM file converted from my ifc file is more than 100MB. The navigation is too bad when using XbimWebUI to open it. In this way, I want to decrease the number of triangles each geometry has to decrease the burden of WebGL. For example, a small circle is represented by 26 triangles now in WebGL. I want to make it consisted by 18 triangles, even if it will look like a bit rough.  

The triangulation part is done in the XbimExchange project. I found precision, deflection and angle are used to write triangulation. Is it possible to decrease the number of triangles by changing some variables or add some more stuff in the relevant function or script like the mesh part of XbimTessellator.cs? If it is, could you please tell me a start point in this project? 

Thank you very much!

Regards,
Ivy

**Comment by martin1cerny** [comment]
Hi @swy0930 , you on the right there are parameters which can decrease complexity of the model. @SteveLockley is the right person to ask about this because he has done all the work on geometry engine and tessellator. My first guess would be to change these before the geometry context is created.
```CS
model.ModelFactors.DeflectionAngle = ???;
model.ModelFactors.DeflectionTolerance = ???;
var context = new Xbim3DModelContext(model);
context.CreateContext(XbimGeometryType.PolyhedronBinary);
```
However, be aware that many complex shapes are triangulated in the authoring software already (like Revit, ArchiCAD, Tekla etc.) and this won't simplify them.

**Comment by swy0930** [comment]
Hi @martin1cerny , thanks for your response. I have found some values of deflection angle and deflection tolerance to decrease the number of triangles, even if I don't exactly know what these two represent in the tessellator part. I just tried some random values and some of them work for me.

Another problem I found is when I debugged on the XbimWebUI project and tried to double clicked on an asset. The entity id returned is not the id of this product. It is not a single case. Double click on many assets of that model returns an entity id 167*****. Obviously, no entity will be found through these number, but I don't know where these number came from. They are not product label, instance label or instance type id. Do you have any idea about this situation?

**Comment by martin1cerny** [comment]
These should always be product labels which is the number after '#' sign in the IFC file, not the GUID. The reason is that ID has to be an integral value so that it can be encoded in an RGBA value for a picking operation in the viewer. 

That also means that if the precision of your GPU is not good enough or for some great values it might go wrong. Unfortunately we didn't find a better way for picking because WebGL doesn't have a built-in support for such operation.

**Comment by swy0930** [comment]
That makes sense. Thank you.

I have got a new question. I found that every time the xbim file created from the same ifc file could have different file size with others. The ones created by cmd.exe and the ones created by Xbim.Exchange.exe from the same ifc file have different sizes. 

Also, if I use the created xbim file to create WexBIM file, the size of xbim file will increase and the size of the WexBIM file is almost doubled, compared with the one created by the ifc file. Then, I use XbimWebUI to compare these two WexBIM files, I found the models were exactly the same, but the big WexBIM uses two shapes to present one product while the small WexBIM only uses one shape to present this product. The big WexBIM doubles the number of shapes used for each product and the redundant shapes cannot be rendered by WebGL, but they do be parsed and stored in GPU.

Is there something wrong with the cache part or maybe something else?

**Comment by martin1cerny** [comment]
This seams like geometry is created twice. What is your `cmd.exe`? I'm not aware of any such an utility in xBIM Toolkit. And which one is bigger? I understand WexBIM issue but it is only consequence of the wrong processing of the original model.

**Comment by swy0930** [comment]
cmd.exe is Command Prompt. When use "start without debugging", it will use cmd.exe to run the script. I am sorry, I made a mistake. The size of xbim file created from ifc and xbim just confused me. I just tested and there is no relevance which one cmd.exe or Xbim.Exchange.exe you choose to run. Sorry for that.
Image 1
![capture](https://cloud.githubusercontent.com/assets/8757175/12431651/8605bf1c-bf5b-11e5-9729-fd4884506f80.PNG)
Image 2
![capture2](https://cloud.githubusercontent.com/assets/8757175/12431664/8f49ccc6-bf5b-11e5-859a-65406b07f04f.PNG)
The first image is created from ifc file and the second one is created from the previous Xbim file in the first image. The size of xbim file itself has changed during the process. Also, the size of wexBIM file is almost doubled.

**Comment by martin1cerny** [comment]
I see. The problem is that `Xbim.Exchange.exe` creates the geometry every time you open the file. But xBIM file from the first screen shot contains geometry already so you effectively create it twice hence the size of wexBIM file is doubled. `Xbim.Exchange.exe` should check if the geometry is there already but why do you do the processing twice? Creating geometry is the most expensive operation and you should definitely avoid doing that multiple times.

**Comment by swy0930** [comment]
I am working on the optimization of wexBIM part. Using xbim file to create wexBIM saves some time of conversion from ifc to xbim, especially for large ifc files. Then, I found this issue where the wexBIM files I created later are much larger than the original one, which confused me. Thank you for your help. Now I am trying to split wexBIM into small pieces sorted by floors and load them separately on WebUI project, since the method of decreasing number of triangles doesn't have a big effect. You are right. Almost all the assets are triangulated in the authoring software already.
